In [1]:
import pandas as pd
import numpy as np
data = pd.read_csv("Churn_Modelling.csv")

In [2]:
data

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [3]:
data.isnull().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64

<h3>as there is no null values the data is cleaned so we can train the model further<br></h3>

In [4]:
data['Gender'] = data['Gender'].map({'Female':0,'Male':1})

In [5]:
print(data['Geography'].unique())

['France' 'Spain' 'Germany']


In [6]:
#as this geography is permitted to just the 3 places lets make one hot encoding
data['Geography'] = data['Geography'].map({'France':0,'Spain':1,'Germany':2})
data['Geography']

0       0
1       1
2       0
3       0
4       1
       ..
9995    0
9996    0
9997    0
9998    2
9999    0
Name: Geography, Length: 10000, dtype: int64

In [8]:
##algorithims using for this project are random forest,logisitic regression and gradient boosting
##we need not to consider surname for our data as it is not required for any kind of manuplations.

In [9]:
x = data.iloc[:,1:-1]
x = x.drop('Surname', axis=1)
x = np.array(x)

In [10]:
y = np.array(data.iloc[:,-1])

In [11]:
#splitting the data:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size = 0.2,random_state = 0)

In [12]:
##random forest implementation
from sklearn.ensemble import RandomForestClassifier

model = RandomForestClassifier(n_estimators = 100)

model.fit(x_train, y_train)

y_pred = model.predict(x_test)

# calculating the training and testing accuracies
print("Training accuracy :", model.score(x_train, y_train))
print("Testing accuracy :", model.score(x_test, y_test))

Training accuracy : 1.0
Testing accuracy : 0.869


<h3>the test accuracy by using random forest is 0.8665<br></h3>

In [13]:
from sklearn.linear_model import LogisticRegression

#creating a model
logreg = LogisticRegression(max_iter=1000)

# Train the model
logreg.fit(x_train, y_train)

# Make predictions on the test set
predictions = logreg.predict(x_test)

# Evaluate the model
test_accuracy = logreg.score(x_test, y_test)
train_accuracy = logreg.score(x_train, y_train)

print('Train Accuracy:',train_accuracy)
print("Test Accuracy :",test_accuracy)

Train Accuracy: 0.796
Test Accuracy : 0.7975


<h3>test accuracy by using Logistic Regression is 0.7975<br></h3>

In [14]:
#gradient boosting implementation
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score
gradient_boost = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)

# Training the Gradient Boosting model
gradient_boost.fit(x_train, y_train)

# Making predictions on the test set
predictions = gradient_boost.predict(x_test)
train_predictions = gradient_boost.predict(x_train)

# Evaluating the model
train_accuracy = accuracy_score(y_train,train_predictions)
test_accuracy = accuracy_score(y_test, predictions)
print(f"Train Accuracy: {train_accuracy}")
print(f"Test Accuracy: {test_accuracy}")

Train Accuracy: 0.873375
Test Accuracy: 0.867


<h3>obtained an accuracy of 0.867 with test data<br></h3>

<h2>USING GRADIENT BOOSTING GIVES US MORE APPROPRIATE CLASSIFICATION AMONG SPECIFIED MODELS<br></h2>